src/Common/src/Microsoft/Win32/SafeHandles/SafeHandleCache.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>Provides a cache for special instances of SafeHandles.</summary>
+    /// <typeparam name="T">Specifies the type of SafeHandle.  The type must have a parameterless constructor.</typeparam>
+    internal static class SafeHandleCache<T> where T : SafeHandle
+    {
+        private static T s_invalidHandle;
+
+        /// <summary>
+        /// Gets a cached, invalid handle.  As the instance is cached, it should either never be Disposed
+        /// or it should override <see cref="SafeHandle.Dispose(bool)"/> to prevent disposal when the
+        /// instance is the <see cref="InvalidHandle"/>.
+        /// </summary>
+        internal static T InvalidHandle
+        {
+            get
+            {
+                T currentHandle = Volatile.Read(ref s_invalidHandle);
+                if (currentHandle == null)
+                {
+                    T newHandle = Activator.CreateInstance<T>();
```

Comment:
Actually, I've just changed it to pass in a delegate to create the instance; that way in the future it can be used even when there's no parameterless ctor at all, regardless of visibility.